### PR TITLE
Translated Refs and The DOM

### PR DESCRIPTION
**Pull request by lunmay**

### CODE REVIEW

File: content/docs/refs-and-the-dom.md [hunk 1]
```diff
@@ -11,33 +11,33 @@ redirect_from:
 permalink: docs/refs-and-the-dom.html
 ---
 
-Refs provide a way to access DOM nodes or React elements created in the render method.
+*Ref* menyediakan cara untuk mengakses simpul DOM atau elemen React yang dibuat dalam *render method*.
 
-In the typical React dataflow, [props](/docs/components-and-props.html) are the only way that parent components interact with their children. To modify a child, you re-render it with new props. However, there are a few cases where you need to imperatively modify a child outside of the typical dataflow. The child to be modified could be an instance of a React component, or it could be a DOM element. For both of these cases, React provides an escape hatch.
+Dalam *dataflow* React yang umum, [*props*](/docs/components-and-props.html) adalah satu-satunya cara bagi komponen induk untuk berinteraksi dengan anaknya. Untuk memodifikasi anak, Anda me-*render* ulang dengan *props* yang baru. Tetapi ada beberapa kasus ketika Anda harus memodifikasi anak secara imperatif di luar *dataflow* yang umum. Anak yang akan dimodifikasi bisa berupa komponen React atau elemen DOM. Pada kedua kasus ini, React menyediakan jalan keluar.
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]: *dataflow* can be translated to aliran data I think 😄

File: content/docs/refs-and-the-dom.md [hunk 2]
```diff
@@ -11,33 +11,33 @@ redirect_from:
...
 
-### When to Use Refs {#when-to-use-refs}
+### Kapan Harus Menggunakan *Ref* {#when-to-use-refs}
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]: Loanwords in the header doesn't have to be italicized 😄

File: content/docs/refs-and-the-dom.md [hunk 3]
```diff
@@ -11,33 +11,33 @@ redirect_from:
...
 
-For example, instead of exposing `open()` and `close()` methods on a `Dialog` component, pass an `isOpen` prop to it.
+Misalnya, alih-alih mengekspos *method* `open()` dan `close()` pada komponen `Dialog`, operkan *props* `isOpen` kepadanya.
 
-### Don't Overuse Refs {#dont-overuse-refs}
+### Jangan Berlebihan Menggunakan *Ref* {#dont-overuse-refs}
 
-Your first inclination may be to use refs to "make things happen" in your app. If this is the case, take a moment and think more critically about where state should be owned in the component hierarchy. Often, it becomes clear that the proper place to "own" that state is at a higher level in the hierarchy. See the [Lifting State Up](/docs/lifting-state-up.html) guide for examples of this.
+Godaan pertama mungkin adalah menggunakan *ref* agar aplikasi "bisa berfungsi". Jika benar demikian, hentikan sejenak dan pikirkan secara kritis, tempat *state* harus berada dalam hierarki komponen. Sering kali nantinya ditemukan bahwa tempat yang lebih baik untuk "memiliki" *state* tersebut adalah di tingkat yang lebih tinggi dalam hierarki. Lihat panduan [Meningkatkan *State*](/docs/lifting-state-up.html) untuk contohnya.
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]: `Lifting State Up` is translated as `Memindahkan State ke Atas` 😄

File: content/docs/refs-and-the-dom.md [hunk 4]
```diff
@@ -51,52 +51,52 @@ class MyComponent extends React.Component {
...
-    // create a ref to store the textInput DOM element
+    // buat ref untuk menyimpan elemen DOM textInput
     this.textInput = React.createRef();
     this.focusTextInput = this.focusTextInput.bind(this);
   }
 
   focusTextInput() {
-    // Explicitly focus the text input using the raw DOM API
-    // Note: we're accessing "current" to get the DOM node
+    // Fokuskan secara eksplisit pada input teks menggunakan API DOM mentah
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]: `raw` here can be translated as `dasar` to give better context I think 😄

File: content/docs/refs-and-the-dom.md [hunk 5]
```diff
@@ -105,11 +105,11 @@ class CustomTextInput extends React.Component {
...
 ```
 
-React will assign the `current` property with the DOM element when the component mounts, and assign it back to `null` when it unmounts. `ref` updates happen before `componentDidMount` or `componentDidUpdate` lifecycle methods.
+React akan menetapkan properti `current` menjadi elemen DOM saat komponen dipasang dan menetapkannya kembali menjadi `null` saat dilepas. Pembaruan `ref` terjadi sebelum *lifecycle method* `componentDidMount` atau `componentDidUpdate`.
 
-#### Adding a Ref to a Class Component {#adding-a-ref-to-a-class-component}
+#### Menambahkan *Ref* ke *Class Component* {#adding-a-ref-to-a-class-component}
 
-If we wanted to wrap the `CustomTextInput` above to simulate it being clicked immediately after mounting, we could use a ref to get access to the custom input and call its `focusTextInput` method manually:
+Jika kita ingin membungkus `CustomTextInput` di atas untuk mensimulasikan peristiwa klik pada masukan teks segera setelah dipasang, kita bisa menggunakan *ref* untuk mendapatkan akses ke input khususnya dan memanggil *method* `focusTextInput` secara manual:
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]: `input` here can be translated as `elemen *input*` to give better context 😄

File: content/docs/refs-and-the-dom.md [hunk 6]
```diff
@@ -153,21 +153,21 @@ class Parent extends React.Component {
     this.textInput = React.createRef();
   }
   render() {
-    // This will *not* work!
+    // Ini tidak akan berfungsi!
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]:
```suggestion
    // Ini *tidak akan* berfungsi!
```

File: content/docs/refs-and-the-dom.md [hunk 7]
```diff
@@ -181,33 +181,33 @@ function CustomTextInput(props) {
...
+Walau Anda bisa [menambahkan *ref* ke *child component*](#adding-a-ref-to-a-class-component), ini bukan solusi yang ideal, karena Anda hanya akan mendapatkan instans komponen, bukan simpul DOM. Selain itu, hal ini tidak akan berfungsi dengan *function component*.
 
-If you use React 16.3 or higher, we recommend to use [ref forwarding](/docs/forwarding-refs.html) for these cases. **Ref forwarding lets components opt into exposing any child component's ref as their own**. You can find a detailed example of how to expose a child's DOM node to a parent component [in the ref forwarding documentation](/docs/forwarding-refs.html#forwarding-refs-to-dom-components).
+Jika Anda menggunakan React 16.3 atau yang lebih baru, kami sarankan untuk menggunakan [*ref forwarding*](/docs/forwarding-refs.html) untuk kasus semacam ini. ***Ref forwarding* memungkinkan komponen untuk memilih untuk mengekspos semua *ref* komponen anak sebagai miliknya sendiri**. Anda bisa menemukan contoh mendetail tentang cara mengekspos simpul DOM anak ke komponen induk [dalam dokumentasi *ref forwarding*](/docs/forwarding-refs.html#forwarding-refs-to-dom-components).
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]:
```suggestion
Jika Anda menggunakan React 16.3 atau yang lebih baru, kami sarankan untuk menggunakan [*ref forwarding*](/docs/forwarding-refs.html) untuk kasus semacam ini. **Ref forwarding* memungkinkan komponen untuk memilih untuk mengekspos semua *ref* komponen anak sebagai miliknya sendiri**. Anda bisa menemukan contoh mendetail tentang cara mengekspos simpul DOM anak ke komponen induk [dalam dokumentasi *ref forwarding*](/docs/forwarding-refs.html#forwarding-refs-to-dom-components).
```

Extra `*` 😄

File: content/docs/refs-and-the-dom.md [hunk 8]
```diff
@@ -181,33 +181,33 @@ function CustomTextInput(props) {
...
+Jika dimungkinkan, kami sarankan untuk tidak mengekspos simpul DOM, walau bisa berguna sebagai jalan keluar. Perhatikan bahwa pendekatan ini membutuhkan tambahan kode pada komponen anak Jika Anda sama sekali tidak memiliki kontrol pada implementasi komponen anak, opsi terakhir Anda adalah menggunakan [`findDOMNode()`](/docs/react-dom.html#finddomnode). Tetapi hal ini tidak disarankan dan akan menjadi *deprecated* dalam [`StrictMode`](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage).
 
-### Callback Refs {#callback-refs}
+### *Callback Ref* {#callback-refs}
 
-React also supports another way to set refs called "callback refs", which gives more fine-grain control over when refs are set and unset.
+React juga mendukung cara lain untuk menetapkan *ref* yang disebut sebagai *"callback ref"*, yang memberikan kontrol lebih mendetail kapan *ref* akan di-*set* dan di-*unset*.
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]: `set` & `unset` can be translated as `pasang` & `lepas` 😄

File: content/docs/refs-and-the-dom.md [hunk 9]
```diff
@@ -221,19 +221,19 @@ class CustomTextInput extends React.Component {
     };
 
     this.focusTextInput = () => {
-      // Focus the text input using the raw DOM API
+      // Fokus pada input teks menggunakan API DOM mentah
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]: `mentah` -> `dasar` 😄

File: content/docs/refs-and-the-dom.md [hunk 10]
```diff
@@ -275,16 +275,16 @@ class Parent extends React.Component {
...
 
-### Legacy API: String Refs {#legacy-api-string-refs}
+### API *Legacy*: *String Ref* {#legacy-api-string-refs}
 
-If you worked with React before, you might be familiar with an older API where the `ref` attribute is a string, like `"textInput"`, and the DOM node is accessed as `this.refs.textInput`. We advise against it because string refs have [some issues](https://github.com/facebook/react/pull/8333#issuecomment-271648615), are considered legacy, and **are likely to be removed in one of the future releases**. 
+Jika sebelumnya Anda bekerja dengan React, Anda mungkin sudah mengenal API lawas dengan atribut `ref` berupa string, misalnya `"textInput"`, dan simpul DOM yang diakses dengan `this.refs.textInput`. Kami sarankan untuk tidak melakukannya karena *string ref* memiliki [beberapa masalah](https://github.com/facebook/react/pull/8333#issuecomment-271648615) dan dipertimbangkan sebagai *legacy*, serta **kemungkinan besar akan dihapus dalam salah satu rilis di masa mendatang**.
```
Comment (regalius) [inline content/docs/refs-and-the-dom.md]: *string*

File: content/docs/refs-and-the-dom.md [hunk 11]
```diff
@@ -11,33 +11,33 @@ redirect_from:
...
-* Integrating with third-party DOM libraries.
+* Mengelola fokus, pemilihan teks, atau pemutaran media.
+* Memicu animasi secara imperatif.
+* Mengintegrasikan dengan *library* DOM pihak ketiga.
 
-Avoid using refs for anything that can be done declaratively.
+Hindari penggunaan *ref* untuk semua yang bisa dilakukan secara deklaratif.
 
-For example, instead of exposing `open()` and `close()` methods on a `Dialog` component, pass an `isOpen` prop to it.
+Misalnya, alih-alih mengekspos *method* `open()` dan `close()` pada komponen `Dialog`, operkan *props* `isOpen` kepadanya.
```
Comment (rizafahmi) [inline content/docs/refs-and-the-dom.md]: operkan props sepertinya agak aneh. Gimana kalau `kirimkan *props*`?

File: content/docs/refs-and-the-dom.md [hunk 12]
```diff
@@ -51,52 +51,52 @@ class MyComponent extends React.Component {
...
 
-The value of the ref differs depending on the type of the node:
+Nilai *ref* berbeda-beda tergantung jenis simpulnya:
 
-- When the `ref` attribute is used on an HTML element, the `ref` created in the constructor with `React.createRef()` receives the underlying DOM element as its `current` property.
-- When the `ref` attribute is used on a custom class component, the `ref` object receives the mounted instance of the component as its `current`.
-- **You may not use the `ref` attribute on function components** because they don't have instances.
+- Ketika atribut `ref` digunakan pada elemen HTML, `ref` yang dibuat dalam konstruktor dengan `React.createRef()` menerima elemen DOM yang mendasari sebagai properti `current` miliknya.
+- Saat atribut `ref` digunakan pada *class component* khusus, objek `ref` menerima instans yang dipasang milik komponen sebagai `current`-nya.
+- **Anda tidak diizinkan untuk menggunakan atribut `ref` pada *function component*** karena komponen tersebut tidak memiliki instans.
```
Comment (rizafahmi) [inline content/docs/refs-and-the-dom.md]: instans? Just use `instance`

File: content/docs/refs-and-the-dom.md [hunk 13]
```diff
@@ -181,33 +181,33 @@ function CustomTextInput(props) {
...
 
-### Callback Refs {#callback-refs}
+### *Callback Ref* {#callback-refs}
 
-React also supports another way to set refs called "callback refs", which gives more fine-grain control over when refs are set and unset.
+React juga mendukung cara lain untuk menetapkan *ref* yang disebut sebagai *"callback ref"*, yang memberikan kontrol lebih mendetail kapan *ref* akan di-*set* dan di-*unset*.
 
-Instead of passing a `ref` attribute created by `createRef()`, you pass a function. The function receives the React component instance or HTML DOM element as its argument, which can be stored and accessed elsewhere. 
+Alih-alih mengoper atribut `ref` yang dibuat oleh `createRef()`, Anda mengoper sebuah fungsi. Fungsi tersebut menerima instans komponen React atau elemen DOM HTML sebagai argumennya, yang bisa disimpan dan diakses di tempat lain.
```
Comment (rizafahmi) [inline content/docs/refs-and-the-dom.md]: `oper` -> `kirim`?

**Pull request merged**